oak-run/src/main/java/org/apache/jackrabbit/oak/run/DataStoreCopyCommand.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.jackrabbit.oak.run;
+
+import joptsimple.OptionParser;
+import joptsimple.OptionSet;
+import joptsimple.OptionSpec;
+import joptsimple.OptionSpecBuilder;
+import org.apache.jackrabbit.oak.commons.IOUtils;
+import org.apache.jackrabbit.oak.run.commons.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.StandardSystemProperty.FILE_SEPARATOR;
+
+/**
+ * Command to concurrently download blobs from an azure datastore using sas token authentication.
+ * <p>
+ * Blobs are stored in a specific folder following the datastore structure format.
+ */
+public class DataStoreCopyCommand implements Command {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DataStoreCopyCommand.class);
+
+    private String sourceRepo;
+    private String includePath;
+    private File fileIncludePath;
+    private String sasToken;
+    private String outDir;
+    private int concurrency;
+
+    @Override
+    public void execute(String... args) throws Exception {
+        parseCommandLineParams(args);
+
+        List<String> ids;
+        if (fileIncludePath != null) {
+            if (fileIncludePath.isDirectory() || !fileIncludePath.exists()) {
+                throw new IllegalArgumentException("file-include-path must be a valid file");
+            }
+            ids = org.apache.commons.io.IOUtils.readLines(Files.newInputStream(fileIncludePath.toPath()), StandardCharsets.UTF_8);
+        } else {
+            ids = Arrays.stream(includePath.split(";")).collect(Collectors.toList());
+        }
+        if (ids.isEmpty()) {
+            throw new IllegalArgumentException("Blob ids must be specified");
+        }
+
+        try (Downloader downloader = new Downloader(concurrency)) {
+            long start = System.currentTimeMillis();
+            List<Downloader.ItemResponse> responses = downloader.download(ids.stream().map(id -> {
```

Comment:
Before calling the download method and starting any download, this code is going through all the files that have to be transferred and preparing them for download. Then once the list is built, the method `download`, it is once again traversing the full list of file to create a new list of `DownloadWorker`. This last traversal is relatively light (just creates a new object), but the first pass done here can be slow, as it can potentially write a log statement. 

Maybe this is not an issue in practice, but I think it would be more efficient to avoid building these intermediate lists, and instead pass the blobs to download one by one to the downloader. This would require a new method `download(Item)`, which would just add a new task to the executor for this blob.

Reply:
Wouldn't that have the downside that you can't show the progress, in number of bytes (remaining)?